Lib/contextlib.py

```diff
@@ -531,10 +531,10 @@ def _fix_exception_context(new_exc, old_exc):
             # Context may not be correct, so find the end of the chain
             while 1:
                 exc_context = new_exc.__context__
-                if exc_context is old_exc:
+                if exc_context is None or exc_context is old_exc:
                     # Context is already set correctly (see issue 20317)
                     return
-                if exc_context is None or exc_context is frame_exc:
+                if exc_context is frame_exc:
```

Comment:
I believe that an exception raised with `__context__` of `None` will already have `__context__` populated before ExitStack receives it.  Therefore this existing handling of `None` was not necessary and not handling the intended case.

When `__context__` is `None` here, it means that it was explicitly set this way at the `raise` site (via a finally block).  So we should not override it to `old_exc`.  This is in line with the behavior of actual nested `with` statements.